util/containers/Dockerfile

```diff
@@ -0,0 +1,86 @@
+# syntax=docker/dockerfile:1.4
+
+ARG NODE_VERSION
+ARG RUST_VERSION
+
+
+FROM --platform=${BUILDPLATFORM} docker.io/node:${NODE_VERSION}-alpine AS frontend-build
+
+WORKDIR /build
+COPY frontend/package.json frontend/package-lock.json /build/
+RUN npm i
+
+COPY frontend .
+RUN npm ci
```

Comment:
I don't understand this part. Why do we first copy the package files and `npm i` when we then copy everything and `npm ci` anyway? To me it seems like the first `COPY` and `RUN npm i` could be removed?

Reply:
In terms of a correct build, yes. This pattern is common to make efficient use of Docker build caches. Every instruction will create a new layer in the current image (note that there can be multiple stages in a build, always starting with a `FROM`). 

By first running `npm i`, this layer will already include all downloaded npm packages. If `packages.json` and `package-lock.json` do not change between builds, Docker will simply use the already built layer containing the packages. Builds will effectively start from the `COPY frontend .` command, and since all packages are already downloaded, `npm ci` will be much faster as well.

Reply:
Thanks for the explanation. See main-thread comment below for more comments.